### ISSUE TITLE
BUG: `f2py` does not respect SIGINT (CTRL-C)

### ISSUE THREAD

**Issue by HaoZeke** [numpy/numpy#20148]
### Describe the issue:

Consider the following Fortran subroutine with an infinite loop. Please note that no one in their right mind would write this, but it illustrates the point.
```fortran
! bla.f90
subroutine main
  do
   print*,"zzz..."
  end do
end subroutine main
```
Now this can be compiled into a nice extension module `f2py -m sloop -c bla.f90` as can be seen in the reproducing code example.

This now ignores all `CTRL-C` and other signals. Note that as implemented, callback functions do handle signals in a sense (i.e. they are mostly `python` which does work with signals and avoid executing `fortran` when a signal is caught).

Handling custom signals set via Python's `signal` library is out of scope for this issue.

### Reproduce the code example:

```python
import sloop
sloop.main()
```

### NumPy/Python version information:

```bash
1.21.2 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:36:15)
[Clang 11.1.0 ]
```

**Comment by seberg** [comment]
As far as I know, there is no solution for this, except checking for interrups regularly using the Python API (and there is no good API exposed to do that if you want to release the GIL right now).

So while correct, this is just how things are. NumPy doesn't check for keyboard interrupts either.  (There used to be code in `fft` that tried to check, but it was horrible and leads to hard crashes if a threaded environment.)

**Comment by seberg** [comment]
@HaoZeke did you try running the fortran code for a _long_ time rather than in an infinite loop?  I assume the Python interrupt handler gets called even if you are "inside fortran".  You just never find out.  That is exactly the same as your sleep example, just there you didn't make the loop infinite.

**Comment by HaoZeke** [comment]
Well for Python `3.10` this seems to do exactly what we need:
```c
static void sigint_handler(int signo){
    PyErr_SetInterrupt();
}
void register_kill_signal_handler_() {
  if (signal(SIGINT, sigint_handler) == SIG_ERR) {
    exit(EXIT_FAILURE);
  }
}
```
![image](https://user-images.githubusercontent.com/4336207/138316967-572338b3-f9ec-4c46-b511-fe0c0937161b.png)

Also:
![image](https://user-images.githubusercontent.com/4336207/138317120-703f5fb8-5868-4940-8f54-e49054f1c831.png)

**Comment by pearu** [comment]
For testing, you can use `usercode` statement to insert custom C code that contains signal handling hooks to generated wrapper functions. See https://numpy.org/devdocs/f2py/signature-file.html#f2py-statements .

If the enhancement is implemented, use the same approach as for `threadsafe` to enable the feature.

**Comment by seberg** [comment]
I agree with Pearu, I don't really understand.  Maybe f2py wrappers can check if an interrupt has occurred while Fortran was running, but it doesn't really seem to matter, since Python checks all the time anyway?

Rather, you want to add API to allow authors to write *interruptable* Fortran code.  And we don't even have API for that in the NumPy C-API?!

This seems all pretty running around in circles, you don't _really_ want to register, this one:
```
  if (signal(SIGINT, SIG_DFL) == SIG_ERR) {
    exit(EXIT_FAILURE);
  }
```
right?  It just kills the whole program!

Look at what `npy_interrupt.h` does:  It stores a longjump point to the end of the execution block (i.e. the slow loop) and then sets a signal handler which does the long jump to the end of that loop!  That works, but means that the long-jump must always point to the correct code (i.e. the one currently being being executed).

I suppose there may be ways to do that in a thread-safe manner, but I don't know.  But we will probably only run in circles until we are clear what you even want here.

Maybe, we can store the long-jump points in a linked-list or thread-local, so that "our" (i.e. the NumPy handler) will clean up correctly and only if it is restoring something sensible there...  But I honestly don't know and presumably thread-local doesn't work because the signal handler runs in a main thread (or its own thread on windows) or so.

**Comment by seberg** [comment]
@pearu, the [FFT](https://github.com/numpy/numpy/issues/5036) [problem](https://github.com/numpy/numpy/issues/11798) is the problem of multiple threads setting up a sigint handler but not being able to restore the correct behaviour cleanly.  It was very easy to observe in old NumPy versions, because even `np.test()` would trigger it reliably.

`NPY_SIGINT_(ON|OFF)` is what the FFT code used to use, and it was unable to clean up correctly in general.  That means that `ctrl+c` crashes the process even _after_ all FFTs finished, just as your example code causes undefined behaviour after the f2py function finishes.  Except `NPY_SIGINT_(ON|OFF)` works fine single threaded, it only starts blowing things up if you ran multiple FFTs threaded.

My whole argument is that I don't think this issue is even actionable.  At least not without spending some more thought on solving the issue of restoring the correct signal handler in a multi-threaded environment (or finding another solution).

Unless you have a use-case where you prefer risking crashing the interpreter (even after your task is complete) to make `ctrl+c` work promptly?  Because my knee jerk reaction is that it is worse than not being able to interrupt the Fortran code using `ctrl+c`.

**Comment by seberg** [comment]
But, yes, if f2py does not use `threading`, then you can add an option to wrap it also into `NPY_SIGINT_ON` and `NPY_SIGINT_OF`, and I think that would be fine.
